File: src/data-scratch-library/dsl/c16_logistic_regression/logistic_regression.py
"""
Logistic regression via gradient ascent.
"""

import logging
import math
from functools import partial, reduce
from typing import List, Tuple

from dsl.c04_linear_algebra.e0401_vectors import dot, vector_add
from dsl.c08_gradient_descent.e0804_minibatch_gd import maximize_batch
from dsl.c08_gradient_descent.e0805_stochastic_gd import maximize_stochastic


def logistic(x: float) -> float:
    """Sigmoid function, clamped for numerical stability."""
    if x > 709:
        return 1.0
    if x < -709:
        return 0.0
    return 1.0 / (1 + math.exp(-x))


def logistic_prime(x: float) -> float:
    """Derivative of the sigmoid function."""
    return logistic(x) * (1 - logistic(x))


def logistic_log_likelihood_i(
    x_i: List[float], y_i: int, beta: List[float]
) -> float:
    """Log-likelihood contribution of observation *i*."""
    if y_i == 1:
        return math.log(logistic(dot(x_i, beta)))
    return math.log(1 - logistic(dot(x_i, beta)))


def logistic_log_likelihood(
    x: List[List[float]], y: List[int], beta: List[float]
) -> float:
    """Total log-likelihood."""
    return sum(
        logistic_log_likelihood_i(x_i, y_i, beta)
        for x_i, y_i in zip(x, y)
    )


def logistic_log_partial_ij(
    x_i: List[float], y_i: int, beta: List[float], j: int
) -> float:
    """Partial derivative of log-likelihood w.r.t. beta[j]."""
    return (y_i - logistic(dot(x_i, beta))) * x_i[j]


def logistic_log_gradient_i(
    x_i: List[float], y_i: int, beta: List[float]
) -> List[float]:
    """Gradient of log-likelihood for observation *i*."""
    return [
        logistic_log_partial_ij(x_i, y_i, beta, j) for j, _ in enumerate(beta)
    ]


def logistic_log_gradient(
    x: List[List[float]], y: List[int], beta: List[float]
) -> List[float]:
    """Total gradient of log-likelihood."""
    return reduce(
        vector_add,
        [logistic_log_gradient_i(x_i, y_i, beta) for x_i, y_i in zip(x, y)],
    )


def score_logistic(
    beta_hat: List[float],
    x_test: List[List[float]],
    y_test: List[int],
) -> Tuple[float, float]:
    """Return (precision, recall) on the test set."""
    tp = fp = tn = fn = 0
    for x_i, y_i in zip(x_test, y_test):
        pred = logistic(dot(beta_hat, x_i))
        if y_i == 1 and pred >= 0.5:
            tp += 1
        elif y_i == 1:
            fn += 1
        elif pred >= 0.5:
            fp += 1
        else:
            tn += 1
    precision = tp / (tp + fp) if (tp + fp) else 0.0
    recall = tp / (tp + fn) if (tp + fn) else 0.0
    return precision, recall


def logistic_fit(
    x: List[List[float]], y: List[int]
) -> List[float]:
    """Fit logistic regression via batch then stochastic gradient ascent."""
    logging.info("fitting logistic regression")
    fn = partial(logistic_log_likelihood, x, y)
    gradient_fn = partial(logistic_log_gradient, x, y)
    beta_0 = [1.0, 1.0, 1.0]
    beta_1 = maximize_batch(fn, gradient_fn, beta_0)
    beta_hat = maximize_stochastic(
        logistic_log_likelihood_i, logistic_log_gradient_i, x, y, beta_1
    )
    logging.info("beta_hat = %s", beta_hat)
    return beta_hat

File: src/data-scratch-cpp-library/dscpp/c16_logistic_regression/logistic_regression.cpp
#include "logistic_regression.h"
#include <cmath>

double logistic(double x) {
    return 1.0 / (1.0 + std::exp(-x));
}

double logistic_prime(double x) {
    return logistic(x) * (1.0 - logistic(x));
}

double logistic_log_likelihood_i(const std::vector<double>& x_i, double y_i, const std::vector<double>& beta) {
    if (y_i == 1.0)
        return std::log(logistic(dot(x_i, beta)));
    else
        return std::log(1.0 - logistic(dot(x_i, beta)));
}

double logistic_log_likelihood(const std::vector<std::vector<double>>& x, const std::vector<double>& y, const std::vector<double>& beta) {
    double total = 0.0;
    for (size_t i = 0; i < x.size(); ++i)
        total += logistic_log_likelihood_i(x[i], y[i], beta);
    return total;
}

std::vector<double> logistic_log_partial_gradient_i(const std::vector<double>& x_i, double y_i, const std::vector<double>& beta) {
    double factor = y_i - logistic(dot(x_i, beta));
    std::vector<double> result(x_i.size());
    for (size_t j = 0; j < x_i.size(); ++j)
        result[j] = factor * x_i[j];
    return result;
}

std::vector<double> logistic_log_gradient(const std::vector<std::vector<double>>& x, const std::vector<double>& y, const std::vector<double>& beta) {
    std::vector<double> result(beta.size(), 0.0);
    for (size_t i = 0; i < x.size(); ++i) {
        auto grad_i = logistic_log_partial_gradient_i(x[i], y[i], beta);
        result = vector_add(result, grad_i);
    }
    return result;
}

LogisticScore score_logistic(const std::vector<double>& beta_hat,
                              const std::vector<std::vector<double>>& x_test,
                              const std::vector<double>& y_test) {
    int tp = 0, fp = 0, fn = 0;
    for (size_t i = 0; i < x_test.size(); ++i) {
        double predict = logistic(dot(beta_hat, x_test[i]));
        if (y_test[i] == 1.0 && predict >= 0.5)      ++tp;
        else if (y_test[i] == 1.0)                    ++fn;
        else if (predict >= 0.5)                      ++fp;
    }
    double precision = (tp + fp > 0) ? static_cast<double>(tp) / (tp + fp) : 0.0;
    double recall    = (tp + fn > 0) ? static_cast<double>(tp) / (tp + fn) : 0.0;
    return {precision, recall};
}

std::vector<double> logistic_fit(const std::vector<std::vector<double>>& x, const std::vector<double>& y) {
    std::vector<double> beta_0 = {1.0, 1.0, 1.0};

    auto fn = [&](const std::vector<double>& beta) {
        return logistic_log_likelihood(x, y, beta);
    };
    auto gradient_fn = [&](const std::vector<double>& beta) {
        return logistic_log_gradient(x, y, beta);
    };

    auto beta_1 = maximize_batch(fn, gradient_fn, beta_0);

    auto target_i = [](const std::vector<double>& x_i, double y_i, const std::vector<double>& beta) {
        return logistic_log_likelihood_i(x_i, y_i, beta);
    };
    auto gradient_i = [](const std::vector<double>& x_i, double y_i, const std::vector<double>& beta) {
        return logistic_log_partial_gradient_i(x_i, y_i, beta);
    };

    return maximize_stochastic(target_i, gradient_i, x, y, beta_1);
}
